buildpacktoml_test.go

```diff
@@ -0,0 +1,729 @@
+									Version: "v1",
+								},
+								Buildpack: lifecycle.Buildpack{ID: "A", Version: "v1"},
+							},
+							{
+								Require: lifecycle.Require{
+									Name:    "some-deprecated-bp-replace-version-dep",
+									Version: "some-version-new",
+								},
+								Buildpack: lifecycle.Buildpack{ID: "A", Version: "v1"},
+							},
+							{
+								Require: lifecycle.Require{
+									Name:     "some-deprecated-bp-add-version-dep",
+									Version:  "v2",
+									Metadata: map[string]interface{}{"version": "v2"},
+								},
+								Buildpack: lifecycle.Buildpack{ID: "A", Version: "v1"},
+							},
+							{
+								Require: lifecycle.Require{
+									Name:     "some-deprecated-bp-move-version-dep",
+									Metadata: map[string]interface{}{"version": "v3"},
+								},
+								Buildpack: lifecycle.Buildpack{ID: "A", Version: "v1"},
+							},
+							{
+								Require: lifecycle.Require{
+									Name:     "some-dep",
+									Metadata: map[string]interface{}{"version": "v1"},
+								},
+								Buildpack: lifecycle.Buildpack{ID: "A", Version: "v1"},
+							},
+							{
+								Require: lifecycle.Require{
+									Name:     "some-replace-version-dep",
+									Metadata: map[string]interface{}{"version": "some-version-new"},
+								},
+								Buildpack: lifecycle.Buildpack{ID: "A", Version: "v1"},
+							},
+						},
+						Labels: []lifecycle.Label{},
+						Met: []string{
+							"some-deprecated-bp-dep",
+							"some-deprecated-bp-replace-version-dep",
+							"some-deprecated-bp-add-version-dep",
+							"some-deprecated-bp-move-version-dep",
+							"some-dep",
+							"some-replace-version-dep",
+						},
+						Processes: []launch.Process{},
+						Slices:    []layers.Slice{},
+					}); s != "" {
+						t.Fatalf("Unexpected:\n%s\n", s)
+					}
+				})
+			})
+
+			when("building fails", func() {
+				it("should error when the output buildpack plan is invalid", func() {
+					env.EXPECT().WithPlatform(platformDir).Return(append(os.Environ(), "TEST_ENV=Av1"), nil)
+					mkfile(t, "bad-key", filepath.Join(appDir, "build-plan-out-A-v1.toml"))
+					if _, err := bpTOML.Build(lifecycle.BuildpackPlan{}, config); err == nil {
+						t.Fatal("Expected error.\n")
+					} else if !strings.Contains(err.Error(), "key") {
+						t.Fatalf("Incorrect error: %s\n", err)
+					}
+				})
+
+				it("should error when top level version and metadata version are both present and do not match", func() {
+					env.EXPECT().WithPlatform(platformDir).Return(append(os.Environ(), "TEST_ENV=Av1"), nil)
+					mkfile(t,
+						"[[entries]]\n"+
+							`name = "dep1"`+"\n"+
+							`version = "v2"`+"\n"+
+							"[entries.metadata]\n"+
+							`version = "v1"`+"\n",
+						filepath.Join(appDir, "build-plan-out-A-v1.toml"),
+					)
+					_, err := bpTOML.Build(lifecycle.BuildpackPlan{}, config)
+					h.AssertNotNil(t, err)
+					expected := "top level version does not match metadata version"
```

Comment:
Note that this conflates two previous errors, whereas before we also had `metadata version does not match top level version`. But I thought that this was worth it to keep platform API logic all in builder.go and do the BOM validation here.